File: src/oop/Punto/AbstractPoint2D.java
package oop.Punto;

import java.util.Objects;

public abstract class AbstractPoint2D implements Point2D{

    @Override
    public boolean equals(Object o){
        if(o == null) return false;
        if(o == this) return true;
        if(o instanceof Point2D p){ //instanceof Point2D
            return getX() == p.getX() && getY() == p.getY();
        }
        return false;
    }

    @Override
    public int hashCode(){
        return Objects.hash(getX(), getY());
    }

    @Override
    public String toString(){
        return "(" + getX() + " , " + getY() + ")";
    }

}

File: Python_Esercizi/Punto/AbstractPoint2D.py
import Punto.Point2D as Point2D
import math

class AbstractPoint2D(Point2D):
    def distanza(self , punto):
        return math.sqrt(math.pow(self.getX() - punto.getX(), 2) + math.pow(self.getY() - punto.getY()))

    def somma(self , punto):
        p = Point2D(self.getX() + punto.getX() , self.getY() + punto.getY())
        return p

    def copy(self):
        return Point2D(self.getX() , self.getY())

    def __eq__(self , other):
        if(other is None):
            return False
        if(self is other):
            return True
        if(isinstance(Point2D , other)):
            return self.getX() == other.getX() and self.getY() == other.getY()
        return False

    def __ne__(self , other):
        return not (self == other)

    def __str__(self):
        return "(" + str(self.getX()) + " ; " + str(self.getY()) + ")"

    def __hash__(self):
        return hash(self.__str__())